server_test.go

```diff
@@ -53,6 +54,27 @@ func TestStopBeforeServe(t *testing.T) {
 	}
 }
 
+func TestGracefulStop(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("failed to create listener: %v", err)
+	}
+
+	server := NewServer()
+	go func() {
+		// make sure Serve() is called
+		time.Sleep(time.Millisecond * 500)
+		server.GracefulStop()
+	}()
+
+	err = server.Serve(lis)
+	if err != nil {
+		t.Fatalf("Serve() returned non-nil error on GracefulStop: %v", err)
```

Comment:
hey @menghanl I only added test for testing `Serve()` should return nil on `GracefulStop()`. 

I found it tricky to actually test the order of `Serve` and `GracefulStop`, the only way I can think of is to make both functions send a message to a shared channel when they return, and compare the order of the messages in the channel, but this requires modifying `Serve` and `GracefulStop` themselves.

Reply:
One way to test this is to make `GracefulStop()` block by having a pending streaming RPC, and check that `Serve()` doesn't return.
We have end2end [tests](https://github.com/grpc/grpc-go/blob/master/test/end2end_test.go#L849) that does similar things.